// @salla.sa/twilight-bundles@0.1.40/dist/vite-plugins/build.js
import { resolve as i } from "path";
import * as o from "fs";
function l() {
  const n = process.cwd(), t = i(n, "src/components");
  if (!o.existsSync(t))
    return {};
  const r = {};
  return o.readdirSync(t).filter((e) => o.statSync(i(t, e)).isDirectory()).forEach((e) => {
    const s = i(t, e, "index.ts");
    o.existsSync(s) && (r[e] = s);
  }), r;
}
function m() {
  return {
    name: "salla-component-build",
    enforce: "pre",
    apply: "build",
    config(n) {
      const t = l();
      return Object.keys(t).length === 0 ? (console.warn("No component entries found in src/components directory"), n) : (console.log("Building components:", Object.keys(t)), {
        ...n,
        build: {
          ...n.build,
          lib: {
            entry: t,
            formats: ["es"],
            fileName: (r, e) => `${e}.js`
          },
          rollupOptions: {
            external: [/^lit/],
            output: {
              globals: {
                lit: "lit",
                "lit-element": "litElement",
                "lit-html": "litHtml"
              }
            }
          }
        }
      });
    }
  };
}
export {
  l as findComponentFiles,
  m as sallaBuildPlugin
};
