// @salla.sa/twilight-bundles@0.1.40/dist/vite-plugins/demo.js
import { findComponentFiles as O } from "./build.js";
import * as i from "fs";
import * as a from "path";
const $ = process.env.TWILIGHT_BUNDLES_URL || "https://cdn.salla.network/js/twilight-bundles/latest/twilight-bundles.js";
let g = process.env.TWILIGHT_FORM_BUILDER_MOCK_BASE_URL || "https://salla.design";
g = g.replace(/\/$/, "") + "/api/v1/form-builder-mock";
function _(e, t) {
  const r = {
    ar: {
      toggleTheme: "تغيير المظهر",
      toggleLang: "English",
      dir: "rtl",
      lang: "ar",
      noSettings: "لا توجد إعدادات لهذا العنصر",
      addSettings: "إضافة إعدادات لهذا العنصر",
      saveSettings: "حفظ التغييرات",
      componentNeedsConfig: "هذا العنصر يحتاج إلى إعدادات ليتم عرضه بشكل صحيح.",
      configureComponent: "إعداد العنصر",
      title: t.twilightBundles.name.ar
    },
    en: {
      toggleTheme: "Toggle Theme",
      toggleLang: "Arabic",
      dir: "ltr",
      lang: "en",
      title: t.twilightBundles.name.en,
      noSettings: "No settings for this component",
      addSettings: "Add settings for this component",
      saveSettings: "Save changes",
      componentNeedsConfig: "This component needs configuration to be displayed properly.",
      configureComponent: "Configure Component"
    }
  };
  return `<!DOCTYPE html>
<html lang="ar" dir="rtl">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>${r.ar.title}</title>
    <script>
      // Retry logic for CSS file
      (function retryLoadTwilightCSS() {
        var src = '/node_modules/@salla.sa/twilight-bundles/dist/twilight-bundles.css';
        var maxRetries = 10;
        var retryDelay = 10000; // 1 second
        var attempts = 0;

        function loadCSS() {
          attempts++;
          
          var link = document.createElement('link');
          link.rel = 'stylesheet';
          link.href = src;
          link.onload = function() {
            console.log('[demo] twilight-bundles.css loaded successfully on attempt ' + attempts);
          };
          link.onerror = function() {
            if (attempts < maxRetries) {
              console.warn('[demo] twilight-bundles.css failed to load (attempt ' + attempts + ') — retrying in ' + retryDelay + 'ms');
              setTimeout(loadCSS, retryDelay);
            } else {
              console.error('[demo] twilight-bundles.css failed to load after ' + maxRetries + ' attempts');
            }
          };
          document.head.appendChild(link);
        }

        loadCSS();
      })();
    <\/script>
    <script>
    localStorage.setItem('FormBuilder::debugger', 1);
    window.customComponents = ${JSON.stringify(e.map((s) => s.url))};
    window.customComponentsSchema = ${JSON.stringify(Object.fromEntries(e.map((s) => [s.name, s.schema])))};
    window.customComponentsRaw = ${JSON.stringify(Object.fromEntries(e.map((s) => [s.name, s.rawComponent])))};
    window.twilightDemoOptions = ${JSON.stringify(t)};
    window.formBuilderMockUrl = '${g}';
    window.translations = ${JSON.stringify(r)};
    window.twilightDemoDynamicComponents = ${JSON.stringify(e)};
    <\/script>
    <link rel="icon" type="image/png" media="(prefers-color-scheme: light)" href="https://cdn.salla.network/images/logo/logo-square.png" />
    <link rel="icon" type="image/png" media="(prefers-color-scheme: dark)" href="https://cdn.salla.network/images/logo/logo-light-square.png" />
    <script type="importmap">
    {
        "imports": {
            "lit": "https://esm.run/lit",
            "lit/": "https://esm.run/lit/"
        }
    }
    <\/script>
    <script type="module" src="https://cdn.salla.network/js/twilight/latest/twilight.esm.js" async><\/script>
    <script type="module" src="${$}" demo-mode defer><\/script>
    <link rel="stylesheet" href="https://cdn.salla.network/fonts/pingarlt.css">
    <link rel="stylesheet" href="https://cdn.salla.network/fonts/sallaicons.css?v=2.0.5">
    
    <style>
      .components-grid {
        display: grid;
        grid-template-columns: ${t.grid.columns};
        gap: ${t.grid.gap};
        margin-top: 0;
        overflow: hidden;
      }

      @media (max-width: ${t.grid.minWidth}) {
        .components-grid {
          grid-template-columns: 1fr;
        }
      }
    </style>
  </head>
  <body>
    <salla-demo-layout></salla-demo-layout>   
    <script>
      // Retry logic for demo script
      (function retryLoadTwilightDemoScript() {
        var src = '/node_modules/@salla.sa/twilight-bundles/dist/vite-plugins/_demo.js';
        var maxRetries = 10;
        var retryDelay = 10000; // 1 second
        var attempts = 0;

        function loadScript() {
          attempts++;
          
          var script = document.createElement('script');
          script.src = src;
          script.onload = function() {
            console.log('[demo] _demo.js loaded successfully on attempt ' + attempts);
          };
          script.onerror = function() {
            if (attempts < maxRetries) {
              console.warn('[demo] _demo.js failed to load (attempt ' + attempts + ') — retrying in ' + retryDelay + 'ms');
              setTimeout(loadScript, retryDelay);
            } else {
              console.error('[demo] _demo.js failed to load after ' + maxRetries + ' attempts');
            }
          };
          document.body.appendChild(script);
        }

        loadScript();
      })();
    <\/script>
    <style>${t.css}</style>
    <script>${t.js}<\/script>
  </body>
</html>`;
}
function b(e) {
  e && i.existsSync(e) && i.unlinkSync(e);
}
function x() {
  const e = a.join(process.cwd(), "twilight-bundle.json");
  return JSON.parse(i.readFileSync(e, "utf-8"));
}
function R(e) {
  try {
    return e != null && e.fields ? (e.fields.push({
      type: "string",
      format: "hidden",
      id: "twilight-bundles-component-name",
      value: e.name
    }), JSON.stringify(e == null ? void 0 : e.fields)) : "";
  } catch (t) {
    return console.error("Error getting schema for component:", t), "";
  }
}
function L(e = {}) {
  let t;
  return {
    name: "salla-component-demo",
    enforce: "pre",
    configResolved(r) {
      var u, f, h, y, w, S;
      if ((u = r.server) != null && u.open)
        return;
      const s = O();
      let C = e.components ? Object.fromEntries(
        Object.entries(s).filter(([n]) => e.components.includes(n))
      ) : s;
      const p = ".salla-temp", l = a.resolve(r.root, "node_modules", p), d = x(), k = Object.entries(C).map(function([n, o]) {
        const T = `/@fs${o}`, v = d.components.find((m) => m.name === n), j = R(v), D = d.components.findIndex((m) => m.name === n);
        return { name: n, path: o, url: T, schema: j, rawComponent: v, order: D };
      }).sort((n, o) => n.order - o.order);
      i.existsSync(l) || i.mkdirSync(l, { recursive: !0 }), t = a.resolve(l, "index.html"), i.writeFileSync(t, _(k, {
        grid: {
          columns: ((f = e.grid) == null ? void 0 : f.columns) || "repeat(1, 1fr)",
          gap: ((h = e.grid) == null ? void 0 : h.gap) || "1rem",
          minWidth: ((y = e.grid) == null ? void 0 : y.minWidth) || "300px"
        },
        css: e.css || "",
        js: e.js || "",
        formbuilder: {
          languages: ((w = e.formbuilder) == null ? void 0 : w.languages) || ["ar", "en"],
          defaultLanguage: ((S = e.formbuilder) == null ? void 0 : S.defaultLanguage) || "ar"
        },
        twilightBundles: d
      }));
      const c = () => {
        b(t);
        try {
          i.rmdirSync(l);
        } catch {
        }
      };
      process.on("SIGINT", c), process.on("SIGTERM", c), process.on("exit", c), r.server.open = `/node_modules/${p}/index.html`;
    },
    configureServer(r) {
    },
    config(r) {
      return {
        ...r,
        server: {
          ...r.server,
          watch: {
            usePolling: !0,
            interval: 100
          }
        }
      };
    },
    closeBundle() {
      b(t);
      const r = a.resolve(process.cwd(), "node_modules/.salla-temp");
      try {
        i.rmdirSync(r);
      } catch {
      }
    }
  };
}
export {
  L as sallaDemoPlugin
};
